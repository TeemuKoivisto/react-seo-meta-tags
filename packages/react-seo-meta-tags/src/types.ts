import * as React from 'react'

export interface ReactSEOMetaTagsProps {
  render?: (el: React.ReactNode) => React.ReactNode
  website?: WebsiteProps
  breadcrumb?: BreadcrumbList
  organization?: OrganizationProps
  blogPost?: BlogPostProps
  facebook?: FacebookProps
  twitter?: TwitterProps
}

/**
 * https://schema.org/WebSite
 */
export interface WebsiteProps {
  /**
   * The URL of this page (eg https://google.com/about)
   */
  url?: string
  /**
   * Maximum 70 characters.
   */
  title: string
  /**
   * The original publication date. ISO 8601 timestamp eg "2019-10-06T13:56:03.123Z"
   * Don't know how useful for random webpages. Add it at least for blog posts.
   */
  datePublished?: string
  /**
   * Maximum 200 characters.
   */
  description?: string
  /**
   * Default "en-US" https://en.wikipedia.org/wiki/IETF_language_tag and https://datahub.io/core/language-codes
   */
  language?: string
  /**
   * URL to the image, PNG, JPEG, or GIF recommended.
   */
  image?: string
  imageAlt?: string
  /**
   * Possibly redundant property. But at least bots can scrape your email and that's fun right? :)
   */
  author?: PersonProps
  site?: {
    /**
     * "If your object is part of a larger web site, the name which should be displayed for the overall site. e.g., "IMDb"."
     * At least Telegram uses this as gray text above the title when sharing links.
     */
    siteName?: string
    /**
     * If your website has a search functionality, enter the URL with parameter here eg "https://www.google.com/search?q="
     */
    searchUrl?: string
  }
}

/**
 * https://developers.facebook.com/docs/sharing/webmasters/
 * http://ogp.me/
 */
export interface FacebookProps {
  /**
   * The title of your article without any branding such as your site name.
   */
  title?: string
  /**
   * A brief description of the content, usually between 2 and 4 sentences.
   */
  description?: string
  /**
   * Used for og:locale. Default "en-US"
   */
  language?: string
  /**
   * Facebook recommends 1200x630 size, ratio of 1.91:1. PNG, JPEG, or GIF.
   * But if you want your image to be displayed as a smaller image (aka thumbnail, similar to 'summary' Twitter card),
   * your image should be smaller than 400x209 and preferably with a ratio of 1:1. Unless your image can be cropped
   * into a 1:1 box without making it look bad. So you probably want to use something like 200x200.
   * Note the "minimum size constraint of 200px by 200px".
   */
  image?: string
  imageAlt?: string
  /**
   * "A URL to a video file that complements this object."
   */
  video?: string
  /**
   * "A URL to an audio file to accompany this object."
   */
  audio?: string
  /**
   * "Insights lets you view analytics for traffic to your site from Facebook."
   */
  facebookAppId?: string
}

/**
 * https://developer.twitter.com/en/docs/tweets/optimize-with-cards/overview/markup.html
 */
export interface TwitterProps {
  /**
   * Title of content (max 70 characters). Fallback: og:title.
   */
  title?: string
  /**
   * Description of content (maximum 200 characters). Fallback: og:description.
   */
  description?: string
  /**
   * Twitter card image, optimal ratio 1.91:1. Recommended: 1200x628. PNG, JPEG, or GIF. Fallback: og:image.
   */
  image?: string
  /**
   * Alt for the image. Fallback: og:image:alt.
   */
  imageAlt?: string
  /**
   * So since I saw that the large image looked dumb if you are using your faceshot as the image, I added the smaller 'summary'
   * type as an option. It should be in 1:1 scale. Also there exists 'player' and 'app' types but since I'm not using those,
   * I'm not going to spend my precious time figuring out how they work for now.
   * @default 'summary_large_image'
   */
  cardType?: 'summary_large_image' | 'summary'
  /**
   * @username of content creator.
   */
  twitterUser?: string
  /**
   * @username of the site eg @nytimes
   */
  twitterSite?: string
}

/**
 * https://schema.org/BlogPosting
 */
export interface BlogPostProps {
  /**
   * The canonical URL for your page. This should be the undecorated URL, without session variables,
   * user identifying parameters, or counters.
   */
  url?: string
  /**
   * Title of the post. Max 70 characters.
   */
  title: string
  description?: string // Should be a short description about the topic, <=200 words. Mainly for SEO purposes.
  image: string // You should add this. Just use the same image as og/twitter eg 1200x630 with 1.91:1 ratio in PNG, JPEG, or GIF.
  imageAlt?: string // Some SEO tools really want this so it's probably wise to add it
  datePublished?: string // The original publication date. Don't change arbitrarily, Google might downrank you.
  dateModified?: string // Google prefers recent content in search results and also users are more likely to click a recent article
  tags?: string[]
  /**
   * Technically either Person or Organization, but since it doesn't make any sense to not to credit this content to a human, use person.
   * From schema.org:
   * "Please note that author is special in that HTML 5 provides a special mechanism for indicating authorship
   * via the rel tag. That is equivalent to this and may be used interchangeably."
   * Which means.. something.
   */
  author?: PersonProps
  /**
   * You should add this since otherwise Google's structured-data tool will complain...
   * However https://webmasters.stackexchange.com/questions/110332/personal-blog-using-structured-data-fails-validation-on-publisher
   */
  publisher?: OrganizationProps
  site?: {
    /**
     * "If your object is part of a larger web site, the name which should be displayed for the overall site. e.g., "IMDb"."
     * Used for og:site_name
     */
    siteName?: string
  }
}

/**
 * https://schema.org/BreadcrumbList
 */
export type BreadcrumbList = BreadcrumbProps[]

export interface BreadcrumbProps {
  url: string
  title: string
  image?: string
}

/**
 * https://schema.org/Person
 */
export interface PersonProps {
  email?: string
  name?: string
  image?: string
}

/**
 * https://schema.org/Organization
 */
export interface OrganizationProps {
  /**
   * @default "Organization"
   * You could use eg "Corporation"
   */
  '@type'?: string
  description?: string
  /**
   * @example "Google"
   */
  name: string
  /**
   * @example "Google Inc"
   */
  legalName?: string
  /**
   *  URL to the logo image.
   */
  logo: string
  /**
   * List of other webpages referencing this organization, eg Wikipedia, Facebook, Twitter, Instagram etc.
   * Will show these sites alongside your website in Google search results and probably boost your SEO rank too.
   */
  sameAs?: string[]
  /**
   * URL to the organization
   * @example "https://abc.xyz"
   */
  url: string
  /**
   * You can nest as many organizations as you'd like, dunno how useful it's
   */
  parentOrganization?: OrganizationProps
}

export type CombinedProps<T> = (WebsiteProps | BlogPostProps) & T
