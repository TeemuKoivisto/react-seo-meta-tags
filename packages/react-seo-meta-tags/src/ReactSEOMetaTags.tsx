import * as React from 'react'

import {
  generateWebsite,
  generateBreadcrumbList,
  generateBlogPosting,
  generateOrganization,
} from './SchemaOrg'

import {
  ReactSEOMetaTagsProps,
  WebsiteProps,
  BlogPostProps,
  FacebookProps,
  TwitterProps,
  CombinedProps,
} from './types'

/**
 * Compact and comprehensive SEO meta tag component for general web pages and blog articles.
 * Includes basic meta tags as well as Facebook and Twitter tags and JSON-LD schema.
 * Originally designed for a Gatsby blog. Can be used in conjunction with react-helmet.
 */
export class ReactSEOMetaTags extends React.PureComponent<ReactSEOMetaTagsProps> {
  /**
   * General tags. OG and Twitter tags both fallback on these three incase description or image is not provided.
   * However by the way this library works, OG and Twitter tags are both extended from the same objects so
   * they'll always be defined or undefined.
   * @param props
   */
  renderGeneral({ title, description, image }: CombinedProps<{}>) {
    return ([
      <title key="title">{ title }</title>,
      description && <meta key="description" name="description" content={description} />,
      image && <meta key="image" name="image" content={image} />
    ])
  }
  renderNonBlogOgTags() {
    return ([
      <meta key="og:type" property="og:type" content="website" />,
    ])
  }
  renderBlogOgTags({ datePublished, dateModified } : BlogPostProps) {
    return ([
      <meta key="og:type" property="og:type" content="article" />,
      datePublished && <meta key="article:published_time" property="article:published_time" content={datePublished} />,
      dateModified && <meta key="article:modified_time" property="article:modified_time" content={dateModified} />,
    ])
  }
  /**
   * https://developers.facebook.com/docs/sharing/webmasters/
   * http://ogp.me/
   * @param props
   */
  renderFacebook(props: CombinedProps<FacebookProps>) {
    const {
      url, title, description, image, imageAlt, video, audio, site, language = 'en-US', facebookAppId
    } = props
    return ([
      url && <meta key="og:url" property="og:url" content={url} />, // Important
      <meta property="og:locale" content={language}/>,
      <meta key="og:title" property="og:title" content={title} />, // Important
      description && <meta key="og:description" property="og:description" content={description} />, // Somewhat important
      // Facebook recommends 1200x630 size, ratio of 1.91:1 but 1200x1200 is also fine
      image && <meta key="og:image" property="og:image" content={image} />, // Important
      imageAlt && <meta key="og:image:alt" property="og:image:alt" content={imageAlt} />, // For visually impaired people
      video && <meta key="og:video" property="og:video" content={video} />,
      audio && <meta key="og:audio" property="og:audio" content={audio} />,
      site && site.siteName && <meta key="og:site_name" property="og:site_name" content={site!.siteName} />, // Eeh... can't hurt?
      facebookAppId && <meta key="fb:app_id" property="fb:app_id" content={facebookAppId}/>
    ])
  }
  renderTwitter(props: CombinedProps<TwitterProps>) {
    const {
      title, description, image, imageAlt, cardType = 'summary_large_image', twitterUser, twitterSite
    } = props
    return ([
      image && <meta key="twitter:card" name="twitter:card" content={cardType} />,
      twitterUser && <meta key="twitter:creator" name="twitter:creator" content={twitterUser} />,
      twitterSite && <meta key="twitter:site" name="twitter:site" content={twitterSite} />,
      <meta key="twitter:title" name="twitter:title" content={title} />,
      description && <meta key="twitter:description" name="twitter:description" content={description} />,
      image && <meta key="twitter:image" name="twitter:image" content={image} />,
      imageAlt && <meta key="twitter:image:alt" name="twitter:image:alt" content={imageAlt} />
    ])
  }
  renderBlogPostSEO(props: ReactSEOMetaTagsProps) {
    const { website, breadcrumb, facebook, twitter, organization } = props
    const blogPost = props.blogPost as BlogPostProps
    return ([
      this.renderGeneral({ ...website, ...blogPost }),
      <script key="application/ld+json" type="application/ld+json">
        {
          JSON.stringify([
            generateWebsite(blogPost),
            breadcrumb && generateBreadcrumbList(breadcrumb),
            generateBlogPosting(blogPost),
            organization && generateOrganization(organization),
          ].filter(obj => obj !== undefined && obj !== null))
        }
      </script>,
      this.renderBlogOgTags(blogPost),
      this.renderFacebook({ ...website, ...blogPost, ...facebook }),
      this.renderTwitter({ ...website, ...blogPost, ...twitter })
    ])
  }
  renderWebsiteSEO(props: ReactSEOMetaTagsProps) {
    const { facebook, breadcrumb, twitter, organization } = props
    const website = props.website as WebsiteProps
    return ([
      this.renderGeneral(website),
      <script key="application/ld+json" type="application/ld+json">
        {
        /**
         * Stringifying eliminates the undefined values, which keeps the JSON-LD somewhat tidy.
         * Some empty objects might remain, but that shouldn't be a problem.
         */
          JSON.stringify([
            generateWebsite(website),
            breadcrumb && generateBreadcrumbList(breadcrumb),
            organization && generateOrganization(organization),
          ].filter(obj => obj !== undefined && obj !== null))
        }
      </script>,
      this.renderNonBlogOgTags(),
      this.renderFacebook({ ...website, ...facebook }),
      this.renderTwitter({ ...website, ...twitter })
    ])
  }
  render() {
    let el: React.ReactNode
    if (this.props.blogPost) {
      el = this.renderBlogPostSEO(this.props)
    } else if (this.props.website) {
      el = this.renderWebsiteSEO(this.props)
    }
    if (this.props.render) {
      return this.props.render(el)
    }
    return el
  }
}
