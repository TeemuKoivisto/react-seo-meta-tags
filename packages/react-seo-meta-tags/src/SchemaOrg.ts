/**
 * JSON-LD is a format for encoding structured data in the Internet.
 * Its popularity is due to the fact Google uses it and advocates websites to use it.
 * There is a lot of different formats for publishing data on JSON-LD but since
 * my own use-case is only for publishing blog posts I'll be using but a small subset
 * of them.
 */

import {
  WebsiteProps,
  BlogPostProps,
  OrganizationProps,
  BreadcrumbList,
  PersonProps,
} from './types'

/**
 * This schema is the bread and butter of websites. It's a simple description which
 * only benefit is I guess that Google might showcase your site in a search result.
 * (if all stars are aligned)
 * @param props
 */
export function generateWebsite({
  author,
  datePublished,
  description,
  language,
  title,
  url,
  image,
  site,
} : WebsiteProps) {
  return {
    '@context': 'http://schema.org',
    '@type': 'WebPage',
    datePublished,
    description,
    image,
    inLanguage: language || 'en-US',
    name: title,
    url,
    author: author && generatePerson(author),
    potentialAction: site && site.searchUrl && {
      '@type': 'SearchAction',
      target: `${site.searchUrl}{search_term_string}`, // Eg "https://www.google.com/search?q=asdf"
      'query-input': 'required name=search_term_string',
    },
  }
}

function generatePerson({
  email,
  name,
  image,
}: PersonProps) {
  return {
    '@type': 'Person',
    name,
    email,
    image,
  }
}

/**
 * Breadcrumb is a nice way to show the structure of your site. People and bots both respect that.
 * The breadcrumb will be showed in the search result directly, so it's important to get it right.
 * I think Google bot tries to do by itself, but I guess it's helpful if you add it by yourself too.
 *
 * https://developers.google.com/search/docs/guides/enhance-site#enable-breadcrumb
 */
export function generateBreadcrumbList(breadcrumbList: BreadcrumbList) {
  return {
    '@context': 'http://schema.org',
    '@type': 'BreadcrumbList',
    itemListElement: breadcrumbList.map((b, i) => (
      {
        '@type': 'ListItem',
        position: i,
        item: {
          '@id': b.url,
          '@type': 'WebPage',
          url: b.url,
          name: b.title,
          image: b.image,
        }
      }
    ))
  }
}

export function generateOrganization(organization: OrganizationProps): Object {
  return {
    '@context': 'http://schema.org',
    '@type': organization['@type'] ? organization['@type'] : 'Organization',
    description: organization.description,
    name: organization.name,
    legalName: organization.legalName,
    sameAs: organization.sameAs,
    url: organization.url,
    logo: organization.logo,
    parentOrganization: organization.parentOrganization && generateOrganization(organization.parentOrganization),
  }
}

/**
 * This is a general BlogPosting schema which covers quite a few attributes.
 * @param props
 */
export function generateBlogPosting({
  url,
  title,
  description,
  image,
  datePublished,
  dateModified,
  tags,
  author,
  publisher,
} : WebsiteProps & BlogPostProps) {
  return {
    '@context': 'http://schema.org',
    '@type': 'BlogPosting',
    url,
    name: title,
    // From https://developers.google.com/search/docs/data-types/article#article_types
    // "The headline of the article. Headlines should not exceed 110 characters."
    headline: title,
    keywords: tags,
    description,
    author: author && generatePerson(author),
    publisher: publisher && generateOrganization(publisher),
    mainEntityOfPage: { // From example markup of JSON-LD https://developers.google.com/search/docs/data-types/article
      '@type': 'WebPage',
      '@id': url, // Indicates that this BlogPosting is the main thing in this URL.
    },
    // From https://developers.google.com/search/docs/data-types/article#article_types
    // "Images should be at least 1200 pixels wide."
    image,
    // thumbnailUrl ?
    datePublished,
    dateModified, // Recommended by https://search.google.com/structured-data/testing-tool
    // Reasoning https://bts.nomadgate.com/medium-evergreen-content
    // Not only does Google prefer to feature more recent content in its search results, but
    // users are also more likely to click an article with a recent date listed next to it.
    // Does it make sense as you can just manipulate the date? Eeeh... Perhaps Google is aware of that.
  }
}
