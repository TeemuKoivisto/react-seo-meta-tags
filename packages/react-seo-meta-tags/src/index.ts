export { ReactSEOMetaTags } from './ReactSEOMetaTags'
export { ReactSEOMetaTags as default } from './ReactSEOMetaTags'
export * from './types'
