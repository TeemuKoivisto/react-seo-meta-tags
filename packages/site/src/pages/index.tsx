import * as React from 'react'
import { graphql, Link } from 'gatsby'
import styled from 'styled-components'

import { Layout } from '../components/Layout'

import { SiteData } from '../types/graphql'

interface Props {
  data: {
    site: SiteData
    allMarkdownRemark: {
      nodes: any[]
    }
  }
  location: any
}

const FrontPage = ({ data, location }: Props) => {
  const posts = data.allMarkdownRemark.nodes
  return (
    <Layout site={data.site}>
      <Container>
        <h1>Hi there!</h1>
        <p>
          This is an example blog to show the functionality of{' '}
          <a href="https://github.com/TeemuKoivisto/react-seo-meta-tags" rel="noopener">
            react-seo-meta-tags
          </a>
          .
        </p>

        <p>You can try the different SEO tools to see what properties it adds.</p>

        <h2>/ (frontpage)</h2>
        <ul>
          <li>
            <a href="https://search.google.com/test/rich-results" rel="noopener">
              Google Rich Results
            </a>
          </li>
          <li>
            <a
              href="https://search.google.com/structured-data/testing-tool#url=https%3A%2F%2Fteemukoivisto.github.io%2Freact-seo-meta-tags"
              rel="noopener"
            >
              Google Structured Data Testing Tool
            </a>
          </li>
          <li>
            <a
              href="https://developers.google.com/speed/pagespeed/insights/?url=https%3A%2F%2Fteemukoivisto.github.io%2Freact-seo-meta-tags"
              rel="noopener"
            >
              Google PageSpeed insights
            </a>
          </li>
          <li>
            <a
              href="https://developers.facebook.com/tools/debug/?q=https%3A%2F%2Fteemukoivisto.github.io%2Freact-seo-meta-tags"
              rel="noopener"
            >
              Facebook Sharing Debugger (requires logging in)
            </a>
          </li>
          <li>
            <a href="https://cards-dev.twitter.com/validator" rel="noopener">
              Twitter Card validator (requires logging in)
            </a>
          </li>
        </ul>

        <h2>/blog/hello-world</h2>
        <ul>
          <li>
            <a href="https://search.google.com/test/rich-results" rel="noopener">
              Google Rich Results
            </a>
          </li>
          <li>
            <a
              href="https://search.google.com/structured-data/testing-tool#url=https%3A%2F%2Fteemukoivisto.github.io%2Freact-seo-meta-tags%2Fblog%2Fhello-world%2F"
              rel="noopener"
            >
              Google Structured Data Testing Tool
            </a>
          </li>
          <li>
            <a
              href="https://developers.google.com/speed/pagespeed/insights/?url=https%3A%2F%2Fteemukoivisto.github.io%2Freact-seo-meta-tags%2Fblog%2Fhello-world"
              rel="noopener"
            >
              Google PageSpeed insights
            </a>
          </li>
          <li>
            <a
              href="https://developers.facebook.com/tools/debug/?q=https%3A%2F%2Fteemukoivisto.github.io%2Freact-seo-meta-tags%2Fblog%2Fhello-world"
              rel="noopener"
            >
              Facebook Sharing Debugger (requires logging in)
            </a>
          </li>
          <li>
            <a href="https://cards-dev.twitter.com/validator" rel="noopener">
              Twitter Card validator (requires logging in)
            </a>
          </li>
        </ul>

        <p>
          The code for this Gatsby site is a bit complicated, but that's sadly how it is with Gatsby
          ðŸ˜¶.
        </p>

        <div>
          <h2>My 5 most recent blog posts</h2>
          <ul>
            {posts.slice(0, 5).map((node: any) => (
              <li key={node.frontmatter.title}>
                <Link to={node.fields.slug}>{node.frontmatter.title}</Link>
              </li>
            ))}
          </ul>
        </div>
      </Container>
    </Layout>
  )
}

export default FrontPage

const Container = styled.div`
  margin-bottom: 3rem;
`

export const query = graphql`
  query FrontPageQuery {
    site {
      ...SiteData
    }
    allMarkdownRemark(sort: { frontmatter: { datePublished: DESC } }) {
      nodes {
        excerpt
        fields {
          slug
        }
        frontmatter {
          datePublished(formatString: "MMMM DD, YYYY")
          title
          description
        }
      }
    }
  }
  fragment SiteData on Site {
    siteMetadata {
      title
      description
      image
      social {
        facebookAppId
        twitter
      }
      imgSiteUrl
      siteUrl
      site {
        siteName
      }
      author {
        name
        image
      }
      organization {
        name
        logo
        url
      }
    }
  }
  fragment BlogPost on MarkdownRemarkEdge {
    node {
      id
      frontmatter {
        title
        description
        datePublished(formatString: "YYYY-MM-DD")
        dateModified(formatString: "YYYY-MM-DD")
        tags
        images {
          publicURL
        }
      }
      fields {
        slug
      }
      excerpt
    }
  }
`
