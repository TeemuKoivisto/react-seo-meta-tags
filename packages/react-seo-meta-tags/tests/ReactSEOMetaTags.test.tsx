import * as React from 'react'
import Helmet, { HelmetTags } from 'react-helmet'
import { mount } from 'enzyme'
import { create } from 'react-test-renderer'

import { ReactSEOMetaTags } from '../src/ReactSEOMetaTags'

import {
  fullWebsiteData,
  fullWebsiteMetaTags,
  fullWebsiteJSONLD,
  fullBlogPost,
  facebookData,
  twitterData,
  breadcrumb
} from './examples'

describe('<ReactSEOMetaTags />', () => {
  test('Renders correctly when placed inside head tag', () => {
    const testRenderer = create(
      <html lang="en">
        <head>
          <meta charSet="utf-8" />
          <ReactSEOMetaTags website={fullWebsiteData} />
        </head>
        <body>
          <main>Body</main>
        </body>
      </html>
    )
    const json = testRenderer.toJSON()
    expect(json).toMatchSnapshot()
  })
  test('Renders correctly with react-helmet', () => {
    // https://github.com/nfl/react-helmet/issues/381#issuecomment-470824081
    // https://github.com/nfl/react-helmet/issues/168#issuecomment-245349049
    mount(
      <div>
        <ReactSEOMetaTags
          render={(el: React.ReactNode) => <Helmet>{el}</Helmet>}
          website={fullWebsiteData}
        />
        <main>Body</main>
      </div>
    )
    const helmet = Helmet.peek()
    // The types are out of sync with this helmet version..
    const { metaTags, scriptTags } = helmet as unknown as HelmetTags
    expect(helmet.title).toEqual(fullWebsiteData.title)
    expect(metaTags).toEqual(fullWebsiteMetaTags)
    const JSONLD = scriptTags.find(t => t.type === 'application/ld+json')
    expect({
      type: JSONLD?.type,
      innerHTML: JSON.parse(JSONLD?.innerHTML || '')
    }).toEqual({
      type: 'application/ld+json',
      innerHTML: fullWebsiteJSONLD
    })
  })
  test('Renders correctly with blog post', () => {
    const testRenderer = create(
      <html lang="en">
        <head>
          <meta charSet="utf-8" />
          <ReactSEOMetaTags website={fullWebsiteData} blogPost={fullBlogPost} />
        </head>
        <body>
          <main>Body</main>
        </body>
      </html>
    )
    const json = testRenderer.toJSON()
    expect(json).toMatchSnapshot()
  })
  test('Renders correctly with all the metadata defined', () => {
    const testRenderer = create(
      <html lang="en">
        <head>
          <meta charSet="utf-8" />
          <ReactSEOMetaTags
            website={fullWebsiteData}
            blogPost={fullBlogPost}
            breadcrumb={breadcrumb}
            organization={{
              name: 'Google',
              legalName: 'Google Inc',
              url: 'https://google.com',
              logo: 'https://google.com/logo.jpg'
            }}
            facebook={facebookData}
            twitter={twitterData}
          />
        </head>
        <body>
          <main>Body</main>
        </body>
      </html>
    )
    const json = testRenderer.toJSON()
    expect(json).toMatchSnapshot()
  })
})
