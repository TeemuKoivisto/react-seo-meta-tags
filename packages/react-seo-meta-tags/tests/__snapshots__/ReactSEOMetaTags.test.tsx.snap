// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ReactSEOMetaTags /> Renders correctly when placed inside head tag 1`] = `
<html
  lang="en"
>
  <head>
    <meta
      charSet="utf-8"
    />
    <title>
      This is a 70 character long title with a lot of padding to make it so!
    </title>
    <meta
      content="This is a 200 character long description of this web page which is quite interesting and which describes its contents well with a lot of relevant keywords and isn't just general marketing mumbo-jumbo."
      name="description"
    />
    <meta
      content="http://website.com/image.png"
      name="image"
    />
    <script
      type="application/ld+json"
    >
      [{"@context":"http://schema.org","@type":"WebPage","datePublished":"2019-10-06T13:56:03.123Z","description":"This is a 200 character long description of this web page which is quite interesting and which describes its contents well with a lot of relevant keywords and isn't just general marketing mumbo-jumbo.","image":"http://website.com/image.png","inLanguage":"en-US","name":"This is a 70 character long title with a lot of padding to make it so!","url":"https://google.com/about","author":{"@type":"Person","name":"John Smith","email":"person@gmail.com","image":"http://john.me/my-face.jpg"},"potentialAction":{"@type":"SearchAction","target":"https://www.google.com/search?q={search_term_string}","query-input":"required name=search_term_string"}}]
    </script>
    <meta
      content="website"
      property="og:type"
    />
    <meta
      content="https://google.com/about"
      property="og:url"
    />
    <meta
      content="en-US"
      property="og:locale"
    />
    <meta
      content="This is a 70 character long title with a lot of padding to make it so!"
      property="og:title"
    />
    <meta
      content="This is a 200 character long description of this web page which is quite interesting and which describes its contents well with a lot of relevant keywords and isn't just general marketing mumbo-jumbo."
      property="og:description"
    />
    <meta
      content="http://website.com/image.png"
      property="og:image"
    />
    <meta
      content="Alt of my website image"
      property="og:image:alt"
    />
    <meta
      content="IMDb"
      property="og:site_name"
    />
    <meta
      content="summary_large_image"
      name="twitter:card"
    />
    <meta
      content="This is a 70 character long title with a lot of padding to make it so!"
      name="twitter:title"
    />
    <meta
      content="This is a 200 character long description of this web page which is quite interesting and which describes its contents well with a lot of relevant keywords and isn't just general marketing mumbo-jumbo."
      name="twitter:description"
    />
    <meta
      content="http://website.com/image.png"
      name="twitter:image"
    />
    <meta
      content="Alt of my website image"
      name="twitter:image:alt"
    />
  </head>
  <body>
    <main>
      Body
    </main>
  </body>
</html>
`;

exports[`<ReactSEOMetaTags /> Renders correctly with all the metadata defined 1`] = `
<html
  lang="en"
>
  <head>
    <meta
      charSet="utf-8"
    />
    <title>
      Is it a bird, is it a plane? Or is it React?
    </title>
    <meta
      content="This post explains everything you need to know about React."
      name="description"
    />
    <meta
      content="https://myblog.xyz/images/react.png"
      name="image"
    />
    <script
      type="application/ld+json"
    >
      [{"@context":"http://schema.org","@type":"WebPage","datePublished":"2020-12-20T02:11:51.893Z","description":"This post explains everything you need to know about React.","image":"https://myblog.xyz/images/react.png","inLanguage":"en-US","name":"Is it a bird, is it a plane? Or is it React?","url":"https://myblog.xyz/articles/what-is-react","author":{"@type":"Person","name":"John Smith","email":"person@gmail.com","image":"http://john.me/my-face.jpg"}},{"@context":"http://schema.org","@type":"BreadcrumbList","itemListElement":[{"@type":"ListItem","position":0,"item":{"@id":"https://example.com/dresses","@type":"WebPage","url":"https://example.com/dresses","name":"Dresses","image":"https://example.com/dresses.jpg"}},{"@type":"ListItem","position":1,"item":{"@id":"https://example.com/dresses-real","@type":"WebPage","url":"https://example.com/dresses-real","name":"Real Dresses","image":"https://example.com/real-dresses.gif"}}]},{"@context":"http://schema.org","@type":"BlogPosting","url":"https://myblog.xyz/articles/what-is-react","name":"Is it a bird, is it a plane? Or is it React?","headline":"Is it a bird, is it a plane? Or is it React?","keywords":["react","javascript","programming","frontend development"],"description":"This post explains everything you need to know about React.","author":{"@type":"Person","name":"John Smith","email":"person@gmail.com","image":"http://john.me/my-face.jpg"},"publisher":{"@context":"http://schema.org","@type":"Corporation","description":"A software engineering consultancy","name":"Bugs and code","legalName":"Bug and code Inc.","sameAs":["https://en.wikipedia.org/wiki/Enron"],"url":"https://bugsandcode.com","logo":"https://bugsandcode.com/logo.png"},"mainEntityOfPage":{"@type":"WebPage","@id":"https://myblog.xyz/articles/what-is-react"},"image":"https://myblog.xyz/images/react.png","datePublished":"2020-12-20T02:11:51.893Z","dateModified":"2020-12-20T02:11:51.893Z"},{"@context":"http://schema.org","@type":"Organization","name":"Google","legalName":"Google Inc","url":"https://google.com","logo":"https://google.com/logo.jpg"}]
    </script>
    <meta
      content="article"
      property="og:type"
    />
    <meta
      content="2020-12-20T02:11:51.893Z"
      property="article:published_time"
    />
    <meta
      content="2020-12-20T02:11:51.893Z"
      property="article:modified_time"
    />
    <meta
      content="https://myblog.xyz/articles/what-is-react"
      property="og:url"
    />
    <meta
      content="en-US"
      property="og:locale"
    />
    <meta
      content="My facebook title"
      property="og:title"
    />
    <meta
      content="My facebook description"
      property="og:description"
    />
    <meta
      content="https://example.com/my-facebook-image.png"
      property="og:image"
    />
    <meta
      content="My facebook image alt"
      property="og:image:alt"
    />
    <meta
      content="https://example.com/my-facebook-video.mpeg"
      property="og:video"
    />
    <meta
      content="https://example.com/my-facebook-audio.mp3"
      property="og:audio"
    />
    <meta
      content="Amazon Inc."
      property="og:site_name"
    />
    <meta
      content="my facebook app id"
      property="fb:app_id"
    />
    <meta
      content="summary"
      name="twitter:card"
    />
    <meta
      content="@michaelcaine"
      name="twitter:creator"
    />
    <meta
      content="@UK"
      name="twitter:site"
    />
    <meta
      content="My twitter title"
      name="twitter:title"
    />
    <meta
      content="My twitter description"
      name="twitter:description"
    />
    <meta
      content="https://example.com/my-twitter-faceshot-image.png"
      name="twitter:image"
    />
    <meta
      content="My twitter image alt"
      name="twitter:image:alt"
    />
  </head>
  <body>
    <main>
      Body
    </main>
  </body>
</html>
`;

exports[`<ReactSEOMetaTags /> Renders correctly with blog post 1`] = `
<html
  lang="en"
>
  <head>
    <meta
      charSet="utf-8"
    />
    <title>
      Is it a bird, is it a plane? Or is it React?
    </title>
    <meta
      content="This post explains everything you need to know about React."
      name="description"
    />
    <meta
      content="https://myblog.xyz/images/react.png"
      name="image"
    />
    <script
      type="application/ld+json"
    >
      [{"@context":"http://schema.org","@type":"WebPage","datePublished":"2020-12-20T02:11:51.893Z","description":"This post explains everything you need to know about React.","image":"https://myblog.xyz/images/react.png","inLanguage":"en-US","name":"Is it a bird, is it a plane? Or is it React?","url":"https://myblog.xyz/articles/what-is-react","author":{"@type":"Person","name":"John Smith","email":"person@gmail.com","image":"http://john.me/my-face.jpg"}},{"@context":"http://schema.org","@type":"BlogPosting","url":"https://myblog.xyz/articles/what-is-react","name":"Is it a bird, is it a plane? Or is it React?","headline":"Is it a bird, is it a plane? Or is it React?","keywords":["react","javascript","programming","frontend development"],"description":"This post explains everything you need to know about React.","author":{"@type":"Person","name":"John Smith","email":"person@gmail.com","image":"http://john.me/my-face.jpg"},"publisher":{"@context":"http://schema.org","@type":"Corporation","description":"A software engineering consultancy","name":"Bugs and code","legalName":"Bug and code Inc.","sameAs":["https://en.wikipedia.org/wiki/Enron"],"url":"https://bugsandcode.com","logo":"https://bugsandcode.com/logo.png"},"mainEntityOfPage":{"@type":"WebPage","@id":"https://myblog.xyz/articles/what-is-react"},"image":"https://myblog.xyz/images/react.png","datePublished":"2020-12-20T02:11:51.893Z","dateModified":"2020-12-20T02:11:51.893Z"}]
    </script>
    <meta
      content="article"
      property="og:type"
    />
    <meta
      content="2020-12-20T02:11:51.893Z"
      property="article:published_time"
    />
    <meta
      content="2020-12-20T02:11:51.893Z"
      property="article:modified_time"
    />
    <meta
      content="https://myblog.xyz/articles/what-is-react"
      property="og:url"
    />
    <meta
      content="en-US"
      property="og:locale"
    />
    <meta
      content="Is it a bird, is it a plane? Or is it React?"
      property="og:title"
    />
    <meta
      content="This post explains everything you need to know about React."
      property="og:description"
    />
    <meta
      content="https://myblog.xyz/images/react.png"
      property="og:image"
    />
    <meta
      content="Alt of my website image"
      property="og:image:alt"
    />
    <meta
      content="Amazon Inc."
      property="og:site_name"
    />
    <meta
      content="this is a fb app id"
      property="fb:app_id"
    />
    <meta
      content="summary_large_image"
      name="twitter:card"
    />
    <meta
      content="Is it a bird, is it a plane? Or is it React?"
      name="twitter:title"
    />
    <meta
      content="This post explains everything you need to know about React."
      name="twitter:description"
    />
    <meta
      content="https://myblog.xyz/images/react.png"
      name="twitter:image"
    />
    <meta
      content="Alt of my website image"
      name="twitter:image:alt"
    />
  </head>
  <body>
    <main>
      Body
    </main>
  </body>
</html>
`;
